### ISSUE TITLE
Invalid character data assertion using clang 13.0.1 with IWYU 0.17

### ISSUE THREAD

**Issue by justincady** [include-what-you-use/include-what-you-use#1016]
Hi,

I'm trying to use IWYU 0.17 with clang 13.0.1. I used clang 13.0.1 to build IWYU. However, when I run it I see this crash:

```
iwyu_lexer_utils.cc:44: Assertion failed: !invalid
```

```
#0  0x00007faea861122b in __GI_raise (sig=sig@entry=6)
#1  0x00007faea86126e8 in __GI_abort ()
#2  0x0000000000d46c2c in include_what_you_use::FatalMessageEmitter::~FatalMessageEmitter() ()
#3  0x00000000011dd814 in include_what_you_use::SourceManagerCharacterDataGetter::GetCharacterData(clang::SourceLocation) const ()
#4  0x00000000012087a8 in include_what_you_use::IwyuPreprocessorInfo::HandlePragmaComment(clang::SourceRange) ()
#5  0x0000000001208729 in include_what_you_use::IwyuPreprocessorInfo::HandleComment(clang::Preprocessor&, clang::SourceRange) ()
#6  0x00000000013f2269 in clang::PPChainedCallbacks::Endif(clang::SourceLocation, clang::SourceLocation) ()
#7  0x00000000013f44c5 in clang::Preprocessor::SkipExcludedConditionalBlock(clang::SourceLocation, clang::SourceLocation, bool, bool, clang::SourceLocation) ()
#8  0x00000000013f748f in clang::Preprocessor::HandleIfdefDirective(clang::Token&, clang::Token const&, bool, bool) ()
#9  0x00000000013f69e5 in clang::Preprocessor::HandleDirective(clang::Token&) ()
#10 0x000000000139ed7d in clang::Lexer::LexTokenInternal(clang::Token&, bool) ()
#11 0x000000000147d164 in clang::Preprocessor::Lex(clang::Token&) ()
#12 0x00000000022b29c6 in clang::ParseAST(clang::Sema&, bool, bool) ()
#13 0x0000000001f9adb7 in clang::FrontendAction::Execute() ()
#14 0x0000000001e97c56 in clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) ()
#15 0x0000000000d3f01c in main ()
```

I can't find any known IWYU issues that look like this. It is happening, as mentioned, in iwyu_lexer_utils.cc:

```
const char* SourceManagerCharacterDataGetter::GetCharacterData(
    SourceLocation loc) const {
  bool invalid;
  const char* data = source_manager_.getCharacterData(loc, &invalid);
  CHECK_(!invalid); // <--------------------
  CHECK_(data);
  return data;
}
```

Some more notes:

- I am able to use clang 10.0.0 and IWYU 0.14 in the same setup without issue
- I am not able to provide reproducing source
- The same issue occurs using clang 13.0.0, so I do not suspect an issue with 13.0.1 in particular

I suspect that something is configured incorrectly and this is not a true defect within IWYU, but I am not sure.

Do you have any idea what system state could trigger that assertion?

Thank you for any advice.

**Comment by kimgr** [comment]
I suspect maybe the source location `loc` is invalid. Given the backtrace, that would indicate that clang calls `HandleComment` with an invalid source range when processing an `#endif`, presumably with a trailing comment.

Maybe you can reduce that from your actual source code?

Overall, it seems like changed behavior in clang, possibly even a bug, but it's hard to tell without a representative input.

**Comment by justincady** [comment]
@kimgr Thank you for taking a look. With the help of a colleague I was able to root cause this, and it was a bug on our side.

Our local LLVM install had a change to `clang/include/clang/Lex/PPCallbacks.h` under an `#ifdef` that I was not aware of. The define was set when building LLVM, but not set when building IWYU...thus we had a library/header mismatch. This explains why `HandleComment` was being called when processing an `#endif` above. Once I corrected the mismatch, things are working properly.

I'm sorry to bother you with this. :(

In the unlikely scenario that someone hits a problem like this and happens upon this issue, check for a mismatch like the one I described above.
